### PR TITLE
Add E0659 for ambiguous names

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Still on the tracks of the "no error without error code" road.

### CODE REVIEW

**Comment by kennytm** [comment]
@bors r-

Travis failed. 

```
[01:28:44] failures:
[01:28:44] 
[01:28:44] ---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0659 (line 11858) stdout ----
[01:28:44] 	error[E0432]: unresolved import `moon`
[01:28:44]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11868:13
[01:28:44]    |
[01:28:44] 12 |     pub use moon::*;
[01:28:44]    |             ^^^^ Maybe a missing `extern crate moon;`?
[01:28:44] 
[01:28:44] error[E0432]: unresolved import `earth`
[01:28:44]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11869:13
[01:28:44]    |
[01:28:44] 13 |     pub use earth::*;
[01:28:44]    |             ^^^^^ Maybe a missing `extern crate earth;`?
[01:28:44] 
[01:28:44] error[E0425]: cannot find function `foo` in module `collider`
[01:28:44]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11872:11
[01:28:44]    |
[01:28:44] 16 | collider::foo(); // ERROR: `foo` is ambiguous
[01:28:44]    |           ^^^ not found in `collider`
[01:28:44] 
[01:28:44] thread 'rustc' panicked at 'Some expected error codes were not found: ["E0659"]', librustdoc/test.rs:294:9
[01:28:44] 
[01:28:44] ---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0659 (line 11884) stdout ----
[01:28:44] 	error[E0432]: unresolved import `moon`
[01:28:44]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11894:13
[01:28:44]    |
[01:28:44] 12 |     pub use moon;
[01:28:44]    |             ^^^^ no `moon` in the root
[01:28:44] 
[01:28:44] error[E0432]: unresolved import `earth`
[01:28:44]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11895:13
[01:28:44]    |
[01:28:44] 13 |     pub use earth;
[01:28:44]    |             ^^^^^ no `earth` in the root
[01:28:44] 
[01:28:44] thread 'rustc' panicked at 'couldn't compile the test', librustdoc/test.rs:289:13
[01:28:44] 
[01:28:44] 
[01:28:44] failures:
[01:28:44]     /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0659 (line 11858)
[01:28:44]     /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0659 (line 11884)
[01:28:44] 
[01:28:44] test result: FAILED. 662 passed; 2 failed; 22 ignored; 0 measured; 0 filtered out```

**Comment by GuillaumeGomez** [comment]
I need to add a main function I think so rustdoc won't wrap them.

**Comment by kennytm** [comment]
The `pub use moon::*;` one is fixed, but the `pub use moon;` one is still failing with other errors.

<details>

```
[01:38:35] ---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0659 (line 11886) stdout ----
[01:38:35] 	error[E0365]: `moon` is private, and cannot be reexported
[01:38:35]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11896:13
[01:38:35]    |
[01:38:35] 11 |     pub use moon;
[01:38:35]    |             ^^^^ reexport of private `moon`
[01:38:35]    |
[01:38:35]    = note: consider declaring type or module `moon` with `pub`
[01:38:35] 
[01:38:35] error[E0365]: `earth` is private, and cannot be reexported
[01:38:35]   --> /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md:11897:13
[01:38:35]    |
[01:38:35] 12 |     pub use earth;
[01:38:35]    |             ^^^^^ reexport of private `earth`
[01:38:35]    |
[01:38:35]    = note: consider declaring type or module `earth` with `pub`
[01:38:35] 
[01:38:35] thread 'rustc' panicked at 'couldn't compile the test', librustdoc/test.rs:289:13
```

</details>

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -1621,6 +1621,59 @@ println!("const value: {}", SomeModule::PRIVATE); // ok!
...
+
+fn main() {
+    collider::foo(); // ERROR: `foo` is ambiguous
+}
+```
+
+This error generally appears when two items with the same name are imported in a
+same module. In here, the `foo` functions are imported and reexported from the
+`collider` module and therefore, when we're using `collider::foo()`, both
+functions collide.
```
Comment (theotherphil) [inline src/librustc_resolve/diagnostics.rs]: Grammar nits: "are imported in a same module" is incorrect. "are imported into a module" is better. "In here" should just be "Here".

**Pull request merged**